pkg/upload/react-native-cocoa.go

```diff
@@ -0,0 +1,195 @@
+package upload
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/bugsnag/bugsnag-cli/pkg/cocoa"
+	"github.com/bugsnag/bugsnag-cli/pkg/log"
+	"github.com/bugsnag/bugsnag-cli/pkg/server"
+	"github.com/bugsnag/bugsnag-cli/pkg/utils"
+)
+
+type ReactNativeCocoa struct {
+	VersionName   string      `help:"The version of the application."`
+	BundleVersion string      `help:"Bundle version for the application. (iOS only)"`
+	Scheme        string      `help:"The name of the scheme to use when building the application."`
+	SourceMap     string      `help:"Path to the source map file" type:"path"`
+	Bundle        string      `help:"Path to the bundle file" type:"path"`
+	Plist         string      `help:"Path to the Info.plist file" type:"path"`
+	Xcworkspace   string      `help:"Path to the .xcworkspace file" type:"path"`
+	CodeBundleID  string      `help:"A unique identifier to identify a code bundle release when using tools like CodePush"`
+	Dev           bool        `help:"Indicates whether the application is a debug or release build"`
+	ProjectRoot   string      `help:"path to remove from the beginning of the filenames in the mapping file" type:"path"`
+	Path          utils.Paths `arg:"" name:"path" help:"Path to directory or file to upload" type:"path" default:"."`
+}
+
+func ProcessReactNativeCocoa(
+	apiKey string,
+	versionName string,
+	bundleVersion string,
+	scheme string,
+	sourceMapPath string,
+	bundlePath string,
+	plistPath string,
+	xcworkspacePath string,
+	codeBundleId string,
+	dev bool,
+	projectRoot string,
+	paths []string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+) error {
+
+	var rootDirPath string
+	var buildSettings *cocoa.XcodeBuildSettings
+
+	for _, path := range paths {
+		// Check/Set the build folder
+		buildDirPath := filepath.Join(path, "ios", "build")
+		rootDirPath = path
+		if !utils.FileExists(buildDirPath) {
+			buildDirPath = filepath.Join(path, "build")
+
+			if utils.FileExists(buildDirPath) {
+				rootDirPath = filepath.Join(path, "..")
+
+			} else if bundlePath == "" || sourceMapPath == "" {
+				return fmt.Errorf("unable to find bundle files or source maps in within " + path)
+			}
+		}
+
+		// Set a default value for projectRoot if it's not defined
+		if projectRoot == "" {
+			projectRoot = rootDirPath
+		}
+
+		// Check if we're missing any of these parameters
+		if bundlePath == "" || plistPath == "" && (apiKey == "" || versionName == "" || bundleVersion == "") {
+
+			// Check to see if we have the xcworkspacePath
+			if xcworkspacePath == "" {
+				// If not, attempt to locate it in the path/ios/ folder
+				if utils.IsDir(filepath.Join(path, "ios")) {
+					workspacePath, err := utils.FindFolderWithSuffix(filepath.Join(path, "ios"), ".xcworkspace")
+					if err != nil {
+						return err
```

Comment:
Yeah I think that works better, so if the workspace file isn't there, this will fail anyway so for all scenarios where we'd error here I should error with something along the lines of "can't find, please specify it etc etc"?

Reply:
If the workspace file isn't there would it necessarily fail? You only need that file to extract information that might have been provided in command line options (as far as I can see); we don't need the file directly for the upload. Workspace and scheme are used to locate the bundle (if a path isn't provided) and plist (which in turn is used to pull version information that might have been provided).

So on L83 we shouldn't error out at all and on L95 and L107, I suggest we only error if the workspace path was provided. Let processing continue and guard use of workspace and scheme name being empty.

It's complicated logic - so perhaps discussing this in person might be beneficial if that doesn't sound right to you.